File: 1-100/1-10/3-longestSubstring/longestSubstring.cpp
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> hashTable;
        int i = 0, startIndex = 0, length = 0;

        for (; i < s.size(); ++i) {
            auto value = hashTable.find(s[i]);
            if (value != hashTable.end() && startIndex <= value->second) {
                length = max(length, i-startIndex);
                startIndex = value->second + 1;
            }
            
            hashTable[s[i]] = i;
        }
        
        return max(length, i - startIndex);
    }
};

File: 1-100/1-10/3-longestSubstring/longestSubstring.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        index = 0
        result = 0
        count = 0
        cutline = -1
        dictionary = {}

        while index < len(s):
            chr = s[index]
            oldIndex = dictionary.get(chr, -1)
            if oldIndex == -1 or cutline >= oldIndex:
                dictionary[chr] = index
                count += 1
            else:
                if result < count:
                    result = count

                count = index - dictionary[chr]
                cutline = dictionary[chr]
                dictionary[chr] = index

            index += 1

        if count > result:
            result = count
        return result

    def lengthOfLongestSubstring2(self, s):
        """
        :type s: str
        :rtype: int
        """

        start = 0
        charset = {}
        max_length = 0

        for end in range(len(s)):
            if charset.get(s[end], -1) < start:
                charset[s[end]] = end
                max_length = max(max_length, end-start+1)
            else:
                start = charset[s[end]] + 1
                charset[s[end]] = end

        return max_length